src/Expression/CaseExpression.php

```diff
@@ -0,0 +1,209 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Db\Expression;
+
+use Yiisoft\Db\Query\QueryInterface;
+use Yiisoft\Db\Schema\Column\ColumnInterface;
+
+use function array_key_exists;
+use function get_object_vars;
+
+/**
+ * Represents a SQL CASE expression.
+ *
+ * A CASE expression allows conditional logic in SQL queries, returning different values based on specified conditions.
+ * It can be used to implement complex logic directly in SQL statements.
+ *
+ * Example usage:
+ *
+ * ```php
+ * $case = (new CaseExpression())
+ *     ->addWhen('condition1', 'result1')
+ *     ->addWhen('condition2', 'result2')
+ *     ->else('defaultResult');
+ * ```
+ *
+ * This will be generated into a SQL CASE expression like:
+ *
+ * ```sql
+ * CASE
+ *     WHEN condition1 THEN 'result1'
+ *     WHEN condition2 THEN 'result2'
+ *     ELSE 'defaultResult'
+ * END
+ * ```
+ *
+ * Example with a specific case value:
+ *
+ * ```php
+ * $case = (new CaseExpression('expression'))
+ *      ->addWhen(1, 'result1')
+ *      ->addWhen(2, 'result2')
+ *      ->else('defaultResult');
+ * ```
+ *
+ * This will be generated into a SQL CASE expression like:
+ *
+ * ```sql
+ * CASE expression
+ *     WHEN 1 THEN 'result1'
+ *     WHEN 2 THEN 'result2'
+ *     ELSE 'defaultResult'
+ * END
+ * ```
+ */
+final class CaseExpression implements ExpressionInterface
+{
+    /**
+     * @var WhenClause[] List of pairs of conditions and their corresponding results in the CASE expression.
+     */
+    private array $whenClauses = [];
+    /**
+     * @var mixed The result to return if no conditions match in the CASE expression.
+     * If not set, the CASE expression will not have an ELSE clause.
+     *
+     * @psalm-suppress PropertyNotSetInConstructor
+     */
+    private mixed $else;
```

Comment:
Move `$else` and `$whenClause` to constructor. This allow to use `CaseExpression` without call setters for cases when all the data is known in advance.